<!DOCTYPE html>
<html lang="en">
  <head>
    <meta charset="UTF-8" />
    <meta name="viewport" content="width=device-width, initial-scale=1.0" />
    <title>Document</title>
  </head>
  <body>
    <script>
      /*
        目标：进一步理解JS中的错误（Error）和错误处理
        1、错误的类型
            Error:所有错误的父类型
            TypeError:数据类型不正确的错误
            ReferenceError:引用的变量不存在
            RangeError:数据值不在其所允许的范围内
            SyntaxError:语法错误
        2、错误处理
            捕获错误：try ... catch
            抛出错误：throw error
        3、错误对象
           message属性：错误相关信息
           stack属性：函数调用栈记录信息
        */
      //常见的内置错误
      // 1、typeError:数据类型不正确的错误
      // console.log(a);  // ReferenceError: a is not defined
      // console.log("b"); //没有捕获error  ,下面的错误不能执行
      // 2、ReferenceError:引用的变量不存在
      // let b=null;
      // console.log(b.xxx);  //TypeError: Cannot read property 'xxx' of null
      // 3、RangeError:数据值不在其所允许的范围内
      // function fn(){
      //     fn()
      // }
      // fn()//RangeError: Maximum call stack size exceeded   注:需要限制范围，所以报错
      // 4、SyntaxError:语法错误
      // const c="""";
      // console.log(c);//SyntaxError: Unexpected string
      //2、错误处理
      // 1、捕获错误：try ... catch、
      // try{
      //   let d;
      //   console.log(d.xxx);
      // }catch(error){
      //   console.log(error);
      //   console.log(error.message);

      // }
      // console.log("出错之后");
      // 2、抛出错误：throw error
      function something() {
        if (Date.now() % 2 === 1) {
          console.log('当前时间为奇数，可以执行任务');
        } else {
          throw new Error('当前时间为偶数，无法进行');
        }
      }
      //捕获处理异常
      try {
        something();
      } catch (error) {
        alert(error.message);
      }
    </script>
  </body>
</html>
